runtime/lua/vim/lsp/sync.lua

```diff
@@ -105,16 +105,13 @@ local function align_end_position(line, byte, offset_encoding)
     char = compute_line_length(line, offset_encoding) + 1
   else
     -- Modifying line, find the nearest utf codepoint
-    local offset = str_utf_end(line, byte)
```

Comment:
Finding the codepoint start and then finding the codepoint end seems a bit redundant... (and involves iterating over the line twice). I'll need to think a bit about this more to suggest a better fix.

Reply:
Not sure there's a way around it here, since `str_utf_end()` doesn't tell you anything about whether you're UTF-aligned. In theory the redundancy should be limited to four bytes at most, unless we're iterating over the entire line in these functions, which is something that should be fixed separately.

Reply:
We are iterating over the entire line in these functions, that's how mb_off/mb_tail work (which these are lua bindings to). AFAIK having the true end codepoint of the character is all you should need to find the end position...

Reply:
After reading the code for `mb_head_off` and `mb_tail_off` it indeed looks like [we're starting the search relative to the offset, not the beginning of the string](https://github.com/neovim/neovim/blob/326e74571be43823ded9fa805a3173bdabda6bec/src/nvim/mbyte.c#L1867). It looks like the only reason we pass in the base pointer is so we know when to stop in the case of malformed UTF-8. So having to do an extra check here should only require at most a couple more memory accesses.

Reply:
I'm almost certain this can be done with one call, it's fine. I can look into this and file the PR myself.